### PR TITLE
Update pkg_resources-stubs for use in pytype_test

### PR DESCRIPTION
**Pull request by Avasam**
Fixes a type issue in pytype_test. https://github.com/python/typeshed/blob/main/tests/pytype_test.py#L158
Unfortunately I can't immediately remove the `type: ignore` comment because typeshed's tests depend on `types-setuptools`.

### CODE REVIEW

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 1]
```diff
@@ -119,10 +125,15 @@ class EntryPoint:
...
-def get_distribution(dist: Requirement | str | Distribution) -> Distribution: ...
+@overload
+def get_distribution(dist: _D) -> _D: ...
+@overload
+def get_distribution(dist: _PkgReqType) -> DistInfoDistribution: ...
 
 class Distribution(IResourceProvider, IMetadataProvider):
-    PKG_INFO: str
+    PKG_INFO: ClassVar[str]
+    # Initialized to None, but is not meant to be instanciated directly
```
Comment (AlexWaygood) [inline stubs/setuptools/pkg_resources/__init__.pyi:135]: comment typo :)

```suggestion
    # Initialized to None, but is not meant to be instantiated directly
```

Comment (Avasam) [reply stubs/setuptools/pkg_resources/__init__.pyi]: That's my french coming through :P

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 2]
```diff
@@ -119,10 +125,15 @@ class EntryPoint:
     def resolve(self) -> Any: ...
 
 def find_distributions(path_item: str, only: bool = ...) -> Generator[Distribution, None, None]: ...
-def get_distribution(dist: Requirement | str | Distribution) -> Distribution: ...
+@overload
+def get_distribution(dist: _D) -> _D: ...
+@overload
+def get_distribution(dist: _PkgReqType) -> DistInfoDistribution: ...
```
Comment (JelleZijlstra) [inline stubs/setuptools/pkg_resources/__init__.pyi:131]: This calls `get_provider()`, which according to our stub below returns a `Distribution`, not a `DistInfoDistribution`. I haven't traced the code to verify which is right, but we should be consistent.

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 3]
```diff
@@ -119,10 +125,15 @@ class EntryPoint:
...
 
 class Distribution(IResourceProvider, IMetadataProvider):
-    PKG_INFO: str
+    PKG_INFO: ClassVar[str]
+    # Initialized to None, but is not meant to be instantiated directly
+    egg_info: str
```
Comment (JelleZijlstra) [inline stubs/setuptools/pkg_resources/__init__.pyi:136]: I don't see this attribute at all (on setuptools 67.3.1).

File: stubs/setuptools/pkg_resources/__init__.pyi (hunk 2 above)
Comment (Avasam) [reply stubs/setuptools/pkg_resources/__init__.pyi:131]: image

Comment (Avasam) [reply stubs/setuptools/pkg_resources/__init__.pyi:131]: I think you're right. Didn't look too far into it but might depend on if the distribution was `.egg`, `.egg-info` or `.dist-info`

File: stubs/setuptools/pkg_resources/__init__.pyi (hunk 3 above)
Comment (Avasam) [reply stubs/setuptools/pkg_resources/__init__.pyi:136]: Found it, it's in `NullProvider`. If you provide no metadata to Distribution, it uses `EmptyProvider` which is a `NullProvider`

```
>>> type(pkg_resources.get_distribution("numpy")._provider)
<class 'pkg_resources.PathMetadata'>
>>> type(Distribution()._provider)
<class 'pkg_resources.EmptyProvider'>
>>> type(get_provider("numpy"))
<class 'pkg_resources.DefaultProvider'>
```
`PathMetadata`, `EmptyProvider` and `DefaultProvider` all have `NullProvider` in common

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 4]
```diff
@@ -243,19 +253,49 @@ def register_loader_type(loader_type: type, provider_factory: Callable[[types.Mo
...
+    module_path: str
     def __init__(self, path: str, egg_info: str) -> None: ...
 
-class ZipProvider(EggProvider): ...
+class ZipProvider(EggProvider):
+    eagers: list[str] | None
+    zip_pre: str
 
 class EggMetadata(ZipProvider, IResourceProvider):
+    loader: types._LoaderProtocol
```
Comment (JelleZijlstra) [inline stubs/setuptools/pkg_resources/__init__.pyi:292]: This comes from the `zipimporter` argument to `__init__` so the types should match.

File: stubs/setuptools/pkg_resources/__init__.pyi [hunk 5]
```diff
@@ -243,19 +253,49 @@ def register_loader_type(loader_type: type, provider_factory: Callable[[types.Mo
...
 
 class IResourceProvider(IMetadataProvider): ...
-class NullProvider: ...
-class EggProvider(NullProvider): ...
+
+class NullProvider:
+    egg_name: str | None
+    egg_info: str | None
+    loader: types._LoaderProtocol | None
+    module_path: str | None
```
Comment (JelleZijlstra) [inline stubs/setuptools/pkg_resources/__init__.pyi:261]:
```suggestion
    module_path: str
```
The initializer unconditionally sets it to a string.

Comment (Avasam) [reply stubs/setuptools/pkg_resources/__init__.pyi:261]: EmptyProvider sets it explicitely to None

Should I keep `module_path: str | None` for sublcassing. Or add `# type: error[assignment]` to `EmptyProvider.module_path` ?

**Review by JelleZijlstra — APPROVED** [review]

**Pull request merged**